File: sdk/client-js/tests/engine.test.ts
import { expect, test, describe } from "bun:test";
import { Evaluator } from "../src/engine";
import type { FlagConfig, UserContext } from "../src/types";

describe("Evaluator", () => {
  const evaluator = new Evaluator();

  test("should return default variation when no rules match", () => {
    const config: FlagConfig = {
      id: "flag-1",
      key: "test-flag",
      enabled: true,
      default_variation_id: "v1",
      variations: [
        { id: "v1", key: "on", value: true },
        { id: "v2", key: "off", value: false },
      ],
      rules: [],
    };
    const context: UserContext = { id: "user-1", attributes: {} };
    const result = evaluator.evaluate(config, context);
    expect(result.value).toBe(true);
    expect(result.reason).toBe("default variation");
  });

  test("should match EQUALS rule", () => {
    const config: FlagConfig = {
      id: "flag-1",
      key: "test-flag",
      enabled: true,
      default_variation_id: "v2",
      variations: [
        { id: "v1", key: "on", value: true },
        { id: "v2", key: "off", value: false },
      ],
      rules: [
        {
          id: "rule-1",
          variation_id: "v1",
          clauses: [
            { attribute: "email", operator: "EQUALS", values: ["test@example.com"] },
          ],
        },
      ],
    };
    const context: UserContext = { id: "user-1", attributes: { email: "test@example.com" } };
    const result = evaluator.evaluate(config, context);
    expect(result.value).toBe(true);
    expect(result.reason).toBe("rule match: rule-1");
  });

  test("should handle percentage rollout consistently", () => {
    const config: FlagConfig = {
      id: "flag-1",
      key: "test-flag",
      enabled: true,
      rollout_percentage: 50,
      rollout_variation_id: "v1",
      default_variation_id: "v2",
      variations: [
        { id: "v1", key: "on", value: true },
        { id: "v2", key: "off", value: false },
      ],
      rules: [],
    };

    // User 1 should be consistent
    const ctx1: UserContext = { id: "user-1", attributes: {} };
    const res1 = evaluator.evaluate(config, ctx1);
    const res1_again = evaluator.evaluate(config, ctx1);
    expect(res1.value).toBe(res1_again.value);

    // Test a few users to ensure we get different results (statistically likely for 50%)
    let onCount = 0;
    for (let i = 0; i < 100; i++) {
      const res = evaluator.evaluate(config, { id: `user-${i}`, attributes: {} });
      if (res.value === true) onCount++;
    }
    // With 100 users and 50% rollout, we expect around 50 'on' results.
    // This is just a basic sanity check for the hashing distribution.
    expect(onCount).toBeGreaterThan(30);
    expect(onCount).toBeLessThan(70);
  });
});

File: internal/engine/engine_test.go
package engine

import (
	"encoding/json"
	"fmt"
	"testing"

	"github.com/google/uuid"
)

func TestEvaluate_Basic(t *testing.T) {
	evaluator := NewEvaluator()

	flagID := uuid.New().String()
	variationID := uuid.New().String()
	variationValue := json.RawMessage(`{"color": "blue"}`)

	config := FlagConfig{
		ID:      flagID,
		Key:     "test-flag",
		Enabled: true,
		DefaultVariationID: &variationID,
		Variations: []Variation{
			{
				ID:    variationID,
				Key:   "variation-1",
				Value: variationValue,
			},
		},
	}

	context := UserContext{
		ID: "user-1",
	}

	result, err := evaluator.Evaluate(config, context)
	if err != nil {
		t.Fatalf("Evaluate failed: %v", err)
	}

	if result.VariationID != variationID {
		t.Errorf("expected variation %s, got %s", variationID, result.VariationID)
	}

	if string(result.Value) != string(variationValue) {
		t.Errorf("expected value %s, got %s", string(variationValue), string(result.Value))
	}
}

func TestEvaluate_Disabled(t *testing.T) {
	evaluator := NewEvaluator()

	flagID := uuid.New().String()
	variationID := uuid.New().String()

	config := FlagConfig{
		ID:      flagID,
		Key:     "test-flag",
		Enabled: false,
		DefaultVariationID: &variationID,
		Variations: []Variation{
			{
				ID:    variationID,
				Key:   "variation-1",
				Value: json.RawMessage(`true`),
			},
		},
	}

	context := UserContext{
		ID: "user-1",
	}

	result, err := evaluator.Evaluate(config, context)
	if err != nil {
		t.Fatalf("Evaluate failed: %v", err)
	}

	if result.Reason != "flag disabled" {
		t.Errorf("expected reason 'flag disabled', got %s", result.Reason)
	}
}

func TestEvaluate_PercentageRollout(t *testing.T) {
	evaluator := NewEvaluator()

	flagID := "flag-1"
	varID_A := "v-a"
	varID_B := "v-b"
	rollout := 50.0

	config := FlagConfig{
		ID:      flagID,
		Key:     "rollout-flag",
		Enabled: true,
		DefaultVariationID: &varID_A,
		RolloutVariationID: &varID_B,
		Variations: []Variation{
			{ID: varID_A, Key: "a", Value: json.RawMessage(`"A"`)},
			{ID: varID_B, Key: "b", Value: json.RawMessage(`"B"`)},
		},
		RolloutPercentage: &rollout,
	}

	// Test a sample of users to check consistency and approximate distribution
	countA := 0
	countB := 0
	total := 1000

	for i := 0; i < total; i++ {
		userID := fmt.Sprintf("user-%d", i)
		result, _ := evaluator.Evaluate(config, UserContext{ID: userID})
		
		if result.VariationID == varID_A {
			countA++
		} else {
			countB++
		}
	}

	if countA < 400 || countA > 600 {
		t.Errorf("Expected roughly 50%% distribution, got A: %d, B: %d", countA, countB)
	}
}

func TestEvaluate_Rules(t *testing.T) {
	evaluator := NewEvaluator()

	flagID := "flag-1"
	varID_Default := "v-default"
	varID_Premium := "v-premium"

	config := FlagConfig{
		ID:      flagID,
		Key:     "premium-flag",
		Enabled: true,
		DefaultVariationID: &varID_Default,
		Variations: []Variation{
			{ID: varID_Default, Key: "default", Value: json.RawMessage(`"regular"`)},
			{ID: varID_Premium, Key: "premium", Value: json.RawMessage(`"premium"`)},
		},
		Rules: []Rule{
			{
				ID:          "premium-users-rule",
				VariationID: varID_Premium,
				Clauses: []Clause{
					{
						Attribute: "plan",
						Operator:  OperatorEquals,
						Values:    []string{"premium"},
					},
				},
			},
		},
	}

	// Test Premium User
	premiumResult, _ := evaluator.Evaluate(config, UserContext{
		ID: "user-premium",
		Attributes: map[string]interface{}{
			"plan": "premium",
		},
	})
	if premiumResult.VariationID != varID_Premium {
		t.Errorf("expected premium variation, got %s", premiumResult.VariationID)
	}

	// Test Regular User
	regularResult, _ := evaluator.Evaluate(config, UserContext{
		ID: "user-regular",
		Attributes: map[string]interface{}{
			"plan": "free",
		},
	})
	if regularResult.VariationID != varID_Default {
		t.Errorf("expected default variation, got %s", regularResult.VariationID)
	}
}
